src/mpi-base.jl

```diff
@@ -5,11 +5,73 @@ typealias MPIDatatype Union{Char,
 typealias MPIBuffertype{T} Union{Ptr{T}, Array{T}, Ref{T}}
 
 # Define a function mpitype(T) that returns the MPI datatype code for
-# a given type T. The dictonary is defined in __init__ so the module
-# can be precompiled
+# a given type T. In the case the the type does not exist, it is created and
+# then returned. The dictonary is defined in __init__ so the module can be
+# precompiled
+
+# accessor and creation function for getting MPI datatypes
+function mpitype{T}(::Type{T})
+
+    if haskey(mpitype_dict, T)  # if the datatype already exists
```

Comment:
OK, so I added a new commit which merges these two function but swaps that `isbit` and `haskey` checks: https://github.com/JuliaParallel/MPI.jl/pull/132/commits/f4721d22b093b344b748915bd840f84be479ffea

From @eschnett  comment I gather that more work is done in this case. @JaredCrean2 If this is incorrect and approach I took before works please let me know and I will revert to that approach.

I think that it is safe to say that we want to user to know that this is what they are doing, i.e., if something isn't `isbit` they should know that pointers are invalid/garbage coming out of the send.